spec.md

```diff
@@ -179,17 +179,155 @@ physically present in the final circuit. Refer to the description of the
+group---statements in a group may not drive components declared outside the
+group.
+
+The circuit below contains one optional group declaration, `Bar`.  Module `Foo`
+contains a group definition that creates a node computed from a port defined in
+the scope of `Foo`.
+
+``` firrtl
+circuit Foo:
+  declgroup Bar:         ; Declaration of group Bar
+
+  module Foo:
+    input a: UInt<1>
+
+    group Bar:           ; Definition of group Bar inside module Foo
+      node notA = not(a)
+
+  extmodule Baz:
+    input a: UInt<1>
+    output b: UInt<1>
+    declgroup Bar:       ; Declaration of group Bar for extmodule Baz
+```
+
+Optional group declarations may be nested.  Optional group declarations are
+declared with the `declgroup`{.firrtl} keyword indented under an existing
+`declgroup`{.firrtl} keyword.  The circuit below contains four optional group
+declarations, three of which are nested.  `Bar` is the top-level group.  `Baz`
+and `Qux` are nested under `Bar`.  `Quz` is nested under `Qux`.
+
+``` firrtl
+circuit Foo:
+  declgroup Bar:
+    declgroup Baz:
+    declgroup Qux:
+      declgroup Quz:
+```
+
+Optional group definitions must match the nesting of declared groups.  Optional
+groups are defined under existing groups with the `group`{.firrtl} keyword
+indented under an existing `group`{.firrtl} keyword.  For the four declared
+groups in the circuit above, the following is a legal nesting of group
+definitions:
+
+``` firrtl
+module Foo:
+  input a: UInt<1>
+
+  group bar of Bar:
+    node notA = not(a)
+    group Baz:
+    group Qux:
+      group Quz:
+        node notNotA = not(notA)
+```
+
+Statements in a nested optional group may only read from ports or declarations
+of the current module, the current group, or a parent group---statements in a
+group may not drive components declared outside the group.  In the above
+example, `notA` is accessible in the group definition of `Quz` because `notA` is
+declared in a parent group.
+
+Group instances may declare output, passive ports of `Probe`{.firrtl} type.
+These ports are accessible in parent modules which instantiate the child.  In
+the example below, module `Baz` defines a group `Bar` with an output port `_a`.
+This port is accessible inside module `Foo` within its defined group `Bar`.
+_Put differently, module `Baz` has an additional port `_a` that is only
+accessible inside a defined group `Bar`_.
+
+``` firrtl
+circuit Foo:
+  declgroup Bar:
+
+  module Baz:
+    wire a: UInt<1>
+
+    group bar of Bar:
+      output _a: Probe<UInt<1>>
+      node notA = not(a)
+      define _a = probe(notA)
+
+  module Foo:
+
+    inst baz of Baz
+
+    group bar of Bar:
+      node _b = baz._a
+```
+
+External modules may also have output, passive ports of probe type declared for
+their optional groups.  The following example shows declared group `Bar` with an
+output port `a`.
+
+``` firrtl
+extmodule Foo:
+  declgroup Bar:
+    output a: Probe<UInt<1>>
+```
+
```

Comment:
It is slightly inconsistent that `declgroup` can have ports when in an external module, but not have ports when in a circuit. Given this, it would be better to use a different keyword here, e.g., `extgroup`, that is specific to declaring a group with ports inside an external module.

Reply:
`hasgroup` ? `usesgroup`?

Reply:
I like `extgroup`. I would also be okay with `group` (makes more sense to me than `declgroup` for this use), but having a different keyword is probably the clearest thing to do.

Reply:
Offline @seldridge mention that there are potential namespace issues with groups defined in extmodules so `group` definitely doesn't make sense, I see how `declgroup` made some sense but yeah I think `extgroup` is probably the best option.